ekf/log.c

```diff
@@ -78,83 +109,169 @@ static void *ekflog_thread(void *args)
-			ekflog_common.head++;
-			return 0;
-		}
+	if (ekflog_common.actBuff->dirty == false) {
+		return 0;
+	}
 
+	if ((ekflog_common.logFlags & EKFLOG_STRICT_MODE) != 0) {
+		/* Waiting for a place to insert logs */
+		do {
+			pthread_cond_wait(&ekflog_common.buffEvent, &ekflog_common.lock);
+		} while (ekflog_common.actBuff->dirty == true);
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+static int ekflog_writeBin(void *msg, size_t msg_len)
+{
+	size_t remainingBuffSize;
+
+	pthread_mutex_lock(&ekflog_common.lock);
+
+	remainingBuffSize = BUFFS_CAPACITY - ekflog_common.actBuff->size;
+
+	if (remainingBuffSize < msg_len) {
+		/* Changing actual buffer for the next one */
+		ekflog_common.actBuff->dirty = true;
+		ekflog_common.actBuff = ekflog_nextBufferGet(ekflog_common.actBuff);
+		ekflog_common.head = ekflog_common.actBuff->buff;
+		pthread_cond_signal(&ekflog_common.buffEvent);
+	}
+
+	if (ekflog_actBuffReadyToWrite() != 0) {
+		/* Dropping the log */
+		ekflog_common.lost++;
+		pthread_mutex_unlock(&ekflog_common.lock);
 		return -1;
 	}
 
-	/* ekflog_common.head >= ekflog_common.tail */
+	memcpy(ekflog_common.head, msg, msg_len);
+	ekflog_common.head += msg_len;
+	ekflog_common.actBuff->size += msg_len;
+
+	pthread_mutex_unlock(&ekflog_common.lock);
+
+	return 0;
+}
+
+
+static int ekflog_writeLogPrefix(char *msg_buf, char msgType, time_t timestamp)
+{
+	uint64_t time = (uint64_t)timestamp;
+
+	ekflog_common.logCnt++;
+
+	memcpy(msg_buf, &ekflog_common.logCnt, sizeof(ekflog_common.logCnt));
+	msg_buf[sizeof(ekflog_common.logCnt)] = msgType;
+	memcpy(&msg_buf[sizeof(ekflog_common.logCnt) + sizeof(msgType)], &time, sizeof(time));
+
+	return sizeof(uint32_t) + sizeof(char) + sizeof(uint64_t);
+}
+
 
-	if (BUFF_LEN - ekflog_common.head - 2 >= MAX_MSG_LEN) {
-		/* Adding new message after previous one */
-		ekflog_common.head++;
+static inline void ekflog_addLogField(char *msg_buf, int *msgEnd, const void *data, size_t dataSize)
+{
+	memcpy(&msg_buf[*msgEnd], data, dataSize);
+	*msgEnd += dataSize;
+}
+
+
+extern int ekflog_timeWrite(time_t timestamp)
+{
+	int msg_len;
+
+	/* Log call with flags that are not enabled is not an error */
+	if ((ekflog_common.logFlags & EKFLOG_TIME) == 0) {
 		return 0;
 	}
 
-	if (ekflog_common.tail - 1 > MAX_MSG_LEN) {
-		/* New message goes to the beginning of buffer */
-		ekflog_common.head = 0;
+	msg_len = ekflog_writeLogPrefix(ekflog_common.msg_buff, 'T', timestamp);
+
+	return ekflog_writeBin(ekflog_common.msg_buff, msg_len);
+}
+
+
+extern int ekflog_senscImuWrite(const sensor_event_t *accEvt, const sensor_event_t *gyrEvt, const sensor_event_t *magEvt)
+{
+	int msg_len;
+
+	/* Log call with flags that are not enabled is not an error */
+	if ((ekflog_common.logFlags & EKFLOG_SENSC) == 0) {
 		return 0;
 	}
 
-	return -1;
+	msg_len = ekflog_writeLogPrefix(ekflog_common.msg_buff, 'I', accEvt->timestamp);
+
+	ekflog_addLogField(ekflog_common.msg_buff, &msg_len, &accEvt->accels.accelX, sizeof(accEvt->accels.accelX));
```

Comment:
If you redesign `ekflog_addLogField()` so that it only accepts `char ptr, const void *data, size_t n` and returns the `n` you can omit multiple data passes in function calls:

```suggestion
	char buff[MAX_MSG_LEN];
	char *msgHead = buff;

	msgHead += ekflog_addLogField(msgHead, &accEvt->accels.accelX, sizeof(accEvt->accels.accelX));
```